conformace/run_e2e.sh: pass shellcheck and autodetect ginkgo nodes

neolit123 opened the pull request:
**What this PR does / why we need it**:
- for ginkgo parallel mode pass -p to autodetect ginkgo nodes.
- disable parallel mode by default (false).
- use provider "skeleton" instead of "local".
- make run_e2e.sh pass shellcheck.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
xref https://github.com/kubernetes/kubernetes/issues/74853

**Special notes for your reviewer**:
NONE

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

/kind bug
/priority important-longterm
/assign @dims @BenTheElder

neolit123 committed:
conformace/run_e2e.sh: pass shellcheck and autodetect ginkgo nodes

- for ginkgo parallel mode pass -p to autodetect ginkgo nodes.
- disable parallel mode by default (false).
- use provider "skeleton" instead of "local".
- make run_e2e.sh pass shellcheck.

spiffxp commented on cluster/images/conformance/run_e2e.sh:
```diff
@@ -39,23 +39,36 @@ saveResults() {
 case ${E2E_PARALLEL} in
-    'y'|'Y')           ginkgo_args+=("--nodes=25") ;;
-    [1-9]|[1-9][0-9]*) ginkgo_args+=("--nodes=${E2E_PARALLEL}") ;;
+    'y'|'Y')           GINKGO_NODES=25 ;;
```
WDYT about this for parity (y or numbers seems weird to me)
```suggestion
    'n'|'N')           GINKGO_NODES=1 ;;
    'y'|'Y')           GINKGO_NODES=25 ;;
```

neolit123 replied:
@dims WDYT?
this reflects in the YAML.

neolit123 replied:
side question here: isn't the default of 25 for parallel a bit too much?

spiffxp commented on cluster/images/conformance/run_e2e.sh:
```diff
@@ -39,23 +39,36 @@ saveResults() {
+
+NUM_NODES="$(kubectl get nodes \
+    -o=jsonpath='{range .items[*]}{.metadata.name}{"\t"}{.spec.taints}{"\n"}{end}' \
+    | grep -cv "node-role.kubernetes.io/master" \
```
I suspect there are other taints you might want (IIRC hybrid windows/linux clusters are supposed to be tainting their windows nodes out) but I won't block for trying this out

neolit123 replied:
copy pasted from @BenTheElder 's code.

neolit123 replied:
i spoke to sig-windows about hybrid and there is a tracking issue somewhere, but not sure if this will land this cycle. last time i've checked there was a blocker in the e2e framework on this front.

neolit123 replied:
@spiffxp `70431`

BenTheElder replied:
My code only needed to handle kind clusters specifically, so we're depending on kubeadm's behavior + our own. Not sure if this is sufficient for all clusters.

neolit123 replied:
i guess that's true. someone might decide to allow scheduling workloads on their CP nodes.
will try to propose something more generic.

spiffxp commented:
/approve
/hold
if you want to address the comments (and I'll let someone else /lgtm)

neolit123 commented:
> Some packages in hack/.shellcheck_failures are passing shellcheck. Please remove them.
>  ./cluster/images/conformance/run_e2e.sh

will do, after we decide on: https://github.com/kubernetes/kubernetes/pull/74854#discussion_r261839715

neolit123 commented:
@dims @spiffxp @BenTheElder 

update:
- remove file from the list of files to ignore shell check
- added proposal of detecting workload-able nodes (waiting feedback on this)

neolit123 commented:
> pull-kubernetes-conformance-image-test Pending — Job triggered.

```
I0303 17:02:59.574]  ---> 2a8f0665106f
I0303 17:02:59.574] Step 10/14 : ENV E2E_PROVIDER="local"
I0303 17:02:59.858]  ---> Running in 7ed189ade0e0
I0303 17:03:00.210] Removing intermediate container 7ed189ade0e0
I0303 17:03:00.210]  ---> 51550c0ca395
I0303 17:03:00.210] Step 11/14 : ENV E2E_PARALLEL="1"
```
huh, this is not running on this PR.

/test pull-kubernetes-conformance-image-test

neolit123 commented:
/skip

timothysc commented on cluster/images/conformance/run_e2e.sh:
```diff
@@ -39,23 +39,32 @@ saveResults() {
+
+# Get the number of nodes to test in the cluster.
+# Any taint with effect:NoSchedule means a node requires special tolerations from deployments.
+NUM_NODES=$(kubectl get nodes -o=jsonpath='{range .items[*]}{.spec.taints}{"\n"}{end}' | grep -cv "effect:NoSchedule")
```
So I like everything here, but let's take out this part and put it in the framework.

timothysc reviewed (changes requested):
minor change request, but everything else +1 

/cc @stevesloka

neolit123 commented:
updated to not include `--num-nodes` detection.
it's to be part of the e2e FW:
https://github.com/kubernetes/kubernetes/pull/74905

updated PR title too.

/test pull-kubernetes-conformance-image-test

neolit123 commented:
/retest

dims commented:
/assign @stevesloka 
/unassign

dims commented:
cc @johnSchnake @brahmaroutu

timothysc reviewed (approved):
/lgtm 
/approve 
/hold 

for final comments ^

stevesloka commented:
/lgtm

timothysc commented:
/hold cancel

waynr commented on cluster/images/conformance/run_e2e.sh:
```diff
@@ -39,23 +39,28 @@ saveResults() {
 
 case ${E2E_PARALLEL} in
-    'y'|'Y')           ginkgo_args+=("--nodes=25") ;;
-    [1-9]|[1-9][0-9]*) ginkgo_args+=("--nodes=${E2E_PARALLEL}") ;;
```
Removing this breaks `sonobuoy`'s `--e2e-parallel` flag which previously let users like me exercise our best judgement in choosing test execution parallelism. Now the only way to increase parallelism for me is to spin up a VM with larger vcpu count.

waynr commented on cluster/images/conformance/run_e2e.sh:
```diff
@@ -39,23 +39,28 @@ saveResults() {
+        # The flag '--p' will automatically detect the optimal number of ginkgo nodes.
+        ginkgo_args+=("--p")
+        # Skip serial tests if parallel mode is enabled.
+        E2E_SKIP="\\[Serial\\]|${E2E_SKIP}" ;;
```
This is somewhat naive. It means any conformance tests marked with `[Serial]` won't be run, which reduces the value of the conformance suite. That being said, I have noticed that the number of such tests has significantly reduced (if i remember correctly, in kubernetes 1.11.x it was around 7).

BenTheElder replied:
You cannot safely run `[Serial]` tagged tests in parallel, that's what the tag means. ginkgo nodes must be zero to run these.

waynr replied:
I think "naive" was probably a bad choice of words, I was actually thinking of "misleading". It's misleading because it allows users to think they have run all conformance tests, when in reality there may be some `[Serial]` tests that haven't been run. Maybe I'm just overreacting though. Anyway, previously for CI purposes I would run `sonobuoy` twice -- once for all the tests that could be run in parallel, and again for all the tests that had to be run serially.

BenTheElder replied:
For submitting conformance results the standard is one run in non-serial mode / a single ginkgo node with a specific config, but doing something like this is definitely very useful for CI results. :+1:

waynr commented on cluster/images/conformance/run_e2e.sh:
```diff
@@ -39,23 +39,28 @@ saveResults() {
+)
+
+set -x
+/usr/local/bin/ginkgo "${ginkgo_args[@]}" /usr/local/bin/e2e.test -- --disable-log-dump --repo-root=/kubernetes --provider="${E2E_PROVIDER}" --report-dir="${RESULTS_DIR}" --kubeconfig="${KUBECONFIG}" | tee "${RESULTS_DIR}"/e2e.log &
```
I suspect this newfangled way of constructing the `ginkgo` command line has consequences for how special characters in `E2E_SKIP` AND `E2E_FOCUS` need to be escaped since the same values for these variables that worked on older kubernetes versions (using the old heptio conformance image) cause my conformance tests to run 0 tests total. This combined with the changes in this PR to the way `E2E_PARALLEL` is handled was sufficient to fool me into thinking that my tests were passing since `sonobuoy e2e <tarball>` only shows the number of failed tests, nothing about the total number of tests run.

neolit123 commented:
@waynr
> Removing this breaks sonobuoy's --e2e-parallel flag which previously let users like me exercise our best judgement in choosing test execution parallelism. Now the only way to increase parallelism for me is to spin up a VM with larger vcpu count.

are you requesting a revert or would you prefer to submit a fix on top of this PR?
cc @johnSchnake

waynr commented:
@neolit123 I think I'll submit a PR with a fix for this if that's okay. I'll try to do that by the end of the day.

neolit123 commented:
sure go ahead.
i'm just wondering why this was reported so late after the PR merged.

waynr commented:
Cause I only just started digging into the different sonobuoy behaviors and ran across this... I thought it might be useful for the folks who originally submitted and reviewed this PR to learn about the downstream impact.

neolit123 commented:
understood, letting the users pick the value makes perfect sense and is the finer grain control, but perhaps we should also allow them to say "please auto-detect this for me".
